R/test-threads.R

```diff
@@ -330,4 +298,63 @@ delete_models <- function(.mods, .tags = "test threads", .force = FALSE){
 }
 
 
+#' Parse model files and overwrite estimation iterations
+#'
+#' @param .mods a bbi model object or list of model objects. Generally created by `test_threads()`.
+#' @param .max_eval Max number of iterations for NONMEM to run.
+#' @param .detect Estimation options that will be set to `.max_eval` if found.
+#'
+#' @importFrom readr parse_number
+#'
+#' @keywords internal
+adjust_estimation_options <- function(.mods, .max_eval, .detect = c("MAXEVAL", "NITER", "NBURN", "MAX")){
+
+  assert_true(all(.detect %in% c("MAXEVAL", "NITER", "NBURN", "MAX")))
+
+  if(!is.null(.max_eval)){
+    assert_int(.max_eval, lower = 1)
+    search_str <- paste(.detect, collapse = "|")
+    map(.mods, function(.mod){
+      mod_path <- get_model_path(.mod)
+      mod_lines <- mod_path %>% readLines() %>% suppressSpecificWarning("incomplete final line found")
+      str_line_loc <- which(grepl(search_str, mod_lines))
+
+      if(is_empty(str_line_loc)){
+        delete_models(.mods = .mods, .force = TRUE) %>% suppressMessages()
+        stop("Neither MAXEVAL or NITER were found in the ctl file. Please ensure one is provided.")
+      }
+
+      str_values <- str_split(mod_lines[str_line_loc], " ")
+      for(i in seq_along(str_values)){
+
+        prevent <- c("MAXEVAL|NITER", "MAXEVAL|MAX", "MAX|NITER") # options that cant be set for the same estimation method
+        for(j in seq_along(prevent)){
+          prevent_loc <- grepl(prevent[j], str_values[[i]])
+          if(length(prevent_loc[prevent_loc]) > 1){
+            delete_models(.mods = .mods, .force = TRUE) %>% suppressMessages()
+            options_drop <- str_split(prevent[j], "\\|")[[1]]
+            stop(glue("Both {options_drop[1]} and {options_drop[2]} were set for the same estimation method. Please ensure only one is set"))
+          }
+        }
+
+        str_loc <- grepl(search_str, str_values[[i]])
+        str_current <- str_values[[i]][str_loc]
+
+        # We dont want to overwrite values that were originally 0
+        if(any(parse_number(str_current) != 0)){
```

Comment:
I'm getting a bit turned around by the code in this `if` block, even when stepping through it interactively. If I'm understanding it right, perhaps you could drop the `if` block entirely and instead change the `str_loc` assignment above from

    str_loc <- grepl(search_str, str_values[[i]])

to

    str_loc <- grepl(search_str, str_values[[i]]) & !grepl("= +0$", str_values[[i]])

Reply:
Ill look into that change. Basically if `NBURN` was set to 0 in the original model, but `.max_eval` is set to 10, I dont want to overwrite `NBURN` and make it 10. Setting that parameter to 0 would be intentional and affect how the model is run, so I want to make sure we dont overwrite cases of 0

Reply:
> Basically if NBURN was set to 0 in the original model [...]

It's not the motivation I was confused by.  It was the code logic.

As for the motivation though, that sounds like another problem that would go away by letting the caller say what they want to replace explicitly: https://github.com/metrumresearchgroup/bbr/pull/542#issuecomment-1213104655

Reply:
> If I understand correctly, the main conceptual problem you're trying to get a handle on is what estimation options `.max_eval` should map to and how those options interact. I wonder if the design of `test_theads` is making that problem for itself. The scientist knows the input model and knows what they would want to replace, so why not leave it up to them to say?
> 
> ```
> test_threads(mod, .threads = c(2, 4),
>              .replace = c(NITER = 500, NBURN = 100))
> ```

Seth and I had actually talked about that solution a couple times. I was totally ok to do it that way, but we decided to not let scientists have that control.

Reply:
> Seth and I had actually talked about that solution a couple times.

I'd recommend revisiting that decision.  It seems to create more problems than it is worth, and, even ignoring that, I think as a user I would find the magical interpretation of `.maxeval` confusing.

Reply:
> > Seth and I had actually talked about that solution a couple times.
> 
> I'd recommend revisiting that decision. It seems to create more problems than it is worth, and, even ignoring that, I think as a user I would find the magical interpretation of `.maxeval` confusing.

Ill defer to @seth127 for that. I personally like the idea of allowing users to have that flexibility, but dont have any strong inclinations